internal/provider/provider.go

```diff
@@ -5,178 +5,183 @@ package provider
-			// This helper will eventually be deprecated after a migration period.
-			project, projDiags := getProjectFromCredentials(ctx, client)
-			if projDiags != nil {
-				if !projDiags.HasError() {
-					diags = append(diags, projDiags...)
-				} else {
-					projDiags = append(projDiags, diag.Errorf("unable to get project from credentials")...)
-					diags = append(diags, projDiags...)
-					return nil, diags
-				}
-			}
+	client, err := clients.NewClient(clients.ClientConfig{
+		ClientID:      clientID,
+		ClientSecret:  clientSecret,
+		SourceChannel: "terraform-provider-hcp",
+	})
 
-			client.Config.OrganizationID = project.Parent.ID
-			client.Config.ProjectID = project.ID
+	if err != nil {
+		resp.Diagnostics.AddError(fmt.Sprintf("unable to create HCP api client: %v", err), "")
+		return
+	}
+
+	projectID := ""
+	if data.ProjectID.ValueString() != "" {
+		projectID = data.ProjectID.ValueString()
+	} else {
+		projectID = os.Getenv("HCP_PROJECT_ID")
+	}
+
+	if projectID != "" {
+		getProjParams := project_service.NewProjectServiceGetParams()
+		getProjParams.ID = projectID
+		project, err := clients.RetryProjectServiceGet(client, getProjParams)
+		if err != nil {
+			resp.Diagnostics.AddError(fmt.Sprintf("unable to fetch project %q: %v", projectID, err), "")
+			return
+		}
+
+		client.Config.ProjectID = project.Payload.Project.ID
+		client.Config.OrganizationID = project.Payload.Project.Parent.ID
+
+	} else {
+		// For the initial release of the HCP TFP, since only one project was allowed per organization at the time,
+		// the provider handled used the single organization's single project by default, instead of requiring the
+		// user to set it. Once multiple projects are available, this helper issues a warning: when multiple projects exist within the org,
+		// a project ID should be set on the provider or on each resource. Otherwise, the oldest project will be used by default.
+		// This helper will eventually be deprecated after a migration period.
+		project, projDiags := getProjectFromCredentialsFramework(ctx, client)
+		if projDiags != nil {
+			if !projDiags.HasError() {
+				resp.Diagnostics.Append(projDiags...)
+			} else {
+				resp.Diagnostics.AddError("unable to get project from credentials", "")
+				return
+			}
 		}
 
-		return client, diags
+		client.Config.OrganizationID = project.Parent.ID
+		client.Config.ProjectID = project.ID
 	}
+
+	var config ProviderFrameworkConfiguration
+	config.Client = client
+	resp.DataSourceData = client
+	resp.ResourceData = client
 }
 
 // getProjectFromCredentials uses the configured client credentials to
 // fetch the associated organization and returns that organization's
 // single project.
-func getProjectFromCredentials(ctx context.Context, client *clients.Client) (project *models.HashicorpCloudResourcemanagerProject, diags diag.Diagnostics) {
+// This differs from the provider.go implementation due to the diagnostics used
+// by the plugin framework.
```

Comment:
We should deduplicate this code, similar to the approach described for `statuspage.go`